### PR TITLE
Kindle Oasis 1,2,3, Scribe: fix startup orientation

### LINKED ISSUE #11269: Kindle with Gyro - starts upside down when launching in U after D (or vice versa)
* KOReader version: v2023.10-55
* Device: Kindle Scribe

#### Issue
Each time the orientation in Kindle menu is changed from `U` to `D` (or vice versa), KOReader starts as upside down. It doesn't hurt anything, because a simple +/- 90 deg rotation is enough to bring it back.

#### Steps to reproduce
1) Start KOReader in `U`.
2) Exit KOReader.
3) Rotate Kindle in hand by 180 deg to go into `D`.
4) Start KOReader.
5) The `Starting KOReader` message shows up upright, but the message about the book being open and KOReader itself then show up upside down.
6) The same is true if you continue exiting/rotating/starting from this orientation, so it is there each time reliably.

##### `crash.log` (if applicable)
I don't yet where to look for.
My suspicion would be it the gyroevent that doesn't account for it?

### PR DESCRIPTION
**Pull request by mergen3107**
Closes #11269
Thank you @NiLuJe !

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/11277)

### CODE REVIEW

**Review by s0mt0chukwu — APPROVED** [review]

**Comment by mergen3107** [comment]
I brought the rotation line outside the `if rotation_mode > 0` block, because when it is `0`, screen doesn't rotate.

Found out when tested exit in `D`, then rotate to `U` and start KOReader.

**Comment by mergen3107** [comment]
@NiLuJe 
In this PR, where I put the rotation logic before the generic init, wouldn't it make sense to remove the `if rotation_mode > 0` block and just put `self.screen.native_rotation_mode = rotation_mode` unconditionally?

**Comment by mergen3107** [comment]
From `framebuffer.lua`:
```
function fb:init()
    self.blitbuffer_rotation_mode = self.bb:getRotation()
    -- asking the framebuffer for orientation is error prone,
    -- so we do this simple heuristic (for now)
    self.screen_size = self:getRawSize()
    if self.screen_size.w > self.screen_size.h and self.is_always_portrait then
        self.screen_size.w, self.screen_size.h = self.screen_size.h, self.screen_size.w
        -- some framebuffers need to be rotated counter-clockwise (they start in landscape mode)
        io.write("FB: Enforcing portrait mode by doing an initial BB rotation\n")
        self.debug("FB: This prevents the use of blitting optimizations. This should instead be fixed on the device's side on startup.")
        self.bb:rotate(-90)
        self.blitbuffer_rotation_mode = self.bb:getRotation()
        assert(not self.forced_rotation, "If forced HW rotation is used, isAlwaysPortrait should not be set.")
    end
    self.native_rotation_mode = self.forced_rotation and self.forced_rotation.default or self.DEVICE_ROTATED_UPRIGHT
    self.cur_rotation_mode = self.native_rotation_mode
end
```
I don't understand this `self.native_rotation_mode = self.forced_rotation and self.forced_rotation.default or self.DEVICE_ROTATED_UPRIGHT`.

Which one ends up being set up? I am confused by syntax.

Unconditional `self.screen.native_rotation_mode = rotation_mode` with `U` results in `rotation_mode` = 0.

But here, are `self.forced_rotation`, `self.forced_rotation.default` zero as well? I know that `self.DEVICE_ROTATED_UPRIGHT` is 0.

**Comment by NiLuJe** [comment]
> self.forced_rotation

That's only used on (very, very) few PB devices. For all intents and purposes, this is always false, and `native_rotation_mode` is almost always set to UR (0).

**Comment by NiLuJe** [comment]
> That's only used on (very, very) few PB devices. For all intents and purposes, this is always false, and `native_rotation_mode` is almost always set to UR (0).

Meaning the current state looks pretty good to me, keeping the `native_rotation_mode` in the != 0 branch, while making the setRota unconditional.

You can *probably* apply the same changes to the Oases, because I doubt anybody really tested that in a while (and even back when people did, there were conflicting reports).

**Comment by mergen3107** [comment]
> You can _probably_ apply the same changes to the Oases, too

OK, I can make a separate PR for that :D 
1) I'll add `awesome` wake/sleeps
2) Kindle:init() moves up
3) rotation stuff

And then someone can test it.

**Comment by NiLuJe** [comment]
It's kind of an all-or-nothing deal, so just do everything at once in this PR. Having the code diverge between devices that ought to behave similarly is a recipe for forgetting about it and terrible support experience down the line ;).

**Comment by NiLuJe** [comment]
See the pile of comments and references in the Oasis 2 block, that should pretty much tell you that the "somebody will test this" step would be a Christmas miracle at this point ;p.

**Comment by mergen3107** [comment]
Oops, sorry, I missed your last two comments (while doing a PR :') )

**Comment by mergen3107** [comment]
Should I merge https://github.com/koreader/koreader/pull/11278 here?

My idea was it would make it easier to track in future what PR was responsible for Oases code specifically :D

**Comment by NiLuJe** [comment]
> I should probably remove `@fixme` words in there, but keep the comments for reference.

Yup, switch @fixme to @note should do the trick. (Hell, this PR might actually *fix it* after all ;p).

**Comment by NiLuJe** [comment]
> Should I merge #11278 here?
> 
> My idea was it would make it easier to track in future what PR was responsible for Oases code specifically :D

Yeah, single PR for all of them please ;).

**Comment by mergen3107** [comment]
Alright, I am on it! :D 
Thanks!

**Comment by mergen3107** [comment]
All done!
Please review.
Before I did https://github.com/koreader/koreader/pull/11277/commits/a64259579c6d275ac12e3dde80b5f7981625a906 CircleCI tests failed. I think these extra buttons in Oases must be declared before `Kindle:init()`, but I am not sure. I'll appreciate any details on that :D

This was the error:
```
[  FAILED  ] 1 test, listed below:
[  FAILED  ] spec/front/unit/device_spec.lua @ 307: device module kindle oasis should interpret orientation event

 1 FAILED TEST
ERROR in finalizer: common/lua-ljsqlite3/init.lua:355: bad argument #1 to 'pairs' (table expected, got nil)
ERROR in finalizer: common/lua-ljsqlite3/init.lua:355: bad argument #1 to 'pairs' (table expected, got nil)
make: *** [Makefile:146: testfront] Error 1

Exited with code exit status 2
```
https://app.circleci.com/pipelines/github/koreader/koreader/9099/workflows/540e6e0a-3ef3-42ff-82cc-25c3b1a27857/jobs/14073

**Comment by NiLuJe** [comment]
Yeah, Generic init needs to come after the custom Input initialization, otherwise you get the Generic input setup ;).

File: frontend/device/kindle/device.lua [hunk 1]
```diff
@@ -918,12 +921,16 @@ function KindleOasis:init()
...
+    Kindle.init(self)
+
+    --- @note See comments in KindleOasis2:init() for details.
     local haslipc, lipc = pcall(require, "liblipclua")
     if haslipc and lipc then
         local lipc_handle = lipc.init("com.github.koreader.screen")
         if lipc_handle then
             local orientation_code = lipc_handle:get_string_property(
                 "com.lab126.winmgr", "accelerometer")
+            logger.dbg("orientation_code = ", orientation_code)
```
Comment (NiLuJe) [inline frontend/device/kindle/device.lua:933]: That's not a bad idea at all ;).

But please add it to *all* the devices (it's missing from the Scribe).

You can also drop the space *after* the equal sign, a comma already ensures whitespace ;).

File: frontend/device/kindle/device.lua [hunk 2]
```diff
@@ -936,17 +943,14 @@ function KindleOasis:init()
                     rotation_mode = self.screen.DEVICE_ROTATED_COUNTER_CLOCKWISE
                 end
             end
-
-            if rotation_mode > 0 then
-                self.screen.native_rotation_mode = rotation_mode
-                self.screen.cur_rotation_mode = rotation_mode
-            end
-
+            logger.dbg("rotation_mode = ", rotation_mode)
```
Comment (NiLuJe) [inline frontend/device/kindle/device.lua:946]: That one is probably unnecessary though, the conversion is pretty straightforward ;).

File: frontend/device/kindle/device.lua (hunk 1 above)
Comment (mergen3107) [reply frontend/device/kindle/device.lua:933]: > That's not a bad idea at all ;).
I am still hoping for people with Oases to come back with testing xD

> it's missing from the Scribe
Lol, overlooked that one, I have it on my Scribe though.

> a comma already ensures whitespace
Oh, that's why! Thanks :D

File: frontend/device/kindle/device.lua (hunk 2 above)
Comment (mergen3107) [reply frontend/device/kindle/device.lua:946]: OK!

File: frontend/device/kindle/device.lua (hunk 1 above)
Comment (mergen3107) [reply frontend/device/kindle/device.lua:933]: > (it's missing from the Scribe)

Hmm, I double-check and that `logger.dbg` is there for Scribe too (as well as for all Oases)! Did you mean something else?

Comment (NiLuJe) [reply frontend/device/kindle/device.lua]: Nope, must be getting cross-eyed, my bad ;p.

**Review by NiLuJe — CHANGES_REQUESTED** [review]
Minor style nits, but looks good otherwise ;).

> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/11277#-:-NmhdqHq61g64UsCQHBw:bs8o3u1)* status: all files reviewed, 2 unresolved discussions (waiting on @mergen3107)

___
*[`frontend/device/kindle/device.lua` line 946 at r4](https://reviewable.io/reviews/koreader/koreader/11277#-Nmhd4E78U1v0DongnTn:-Nmhd4E78U1v0DongnTo:b-c42ofh) ([raw file](https://github.com/koreader/koreader/blob/1603dc219f672d04e3c69f378c3b012381b34309/frontend/device/kindle/device.lua#L946)):*
> ```Lua
>                 end
>             end
>             if rotation_mode > 0 then self.screen.native_rotation_mode = rotation_mode end
> ```

Not a fan of the one-liners. We're in 2023, whitespace is cheap and greatly enhances readability :).

(For... anything, really, I barely tolerate the `if conditional then return end` ones... ^^).

(So, consider this a style nit for *all* your numerous one-liners in this PR ;p).

**Comment by mergen3107** [comment]
> (So, consider this a style nit for _all_ your numerous one-liners in this PR ;p).

This makes it 8 :D OK,  I can revert them back.
I got influenced by recent
https://github.com/koreader/koreader/blob/4c2fc1eb817528fb02f3d7b5583cfd7c98c197c4/frontend/util.lua#L757-L759

**Comment by mergen3107** [comment]
All done!

**Review by NiLuJe — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/11277)* status: all files reviewed, 2 unresolved discussions (waiting on @mergen3107)

___
*[`frontend/device/kindle/device.lua` line 948 at r5](https://reviewable.io/reviews/koreader/koreader/11277#-NmhjPd35dy_aw86E2lm:-NmhjPd35dy_aw86E2ln:b-nyeanu) ([raw file](https://github.com/koreader/koreader/blob/2528cd45d87c17fe2026826b989585e1a15c4ceb/frontend/device/kindle/device.lua#L948)):*
> ```Lua
>                 end
>             end
>             if rotation_mode > 0 then self.screen.native_rotation_mode = rotation_mode end
> ```

Missed those ;p.

**Comment by mergen3107** [comment]
Thanks!
Is this fine without newlines around the block 
```
            if rotation_mode > 0 then
                self.screen.native_rotation_mode = rotation_mode
            end
````
?

**Review by NiLuJe — APPROVED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/11277)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @mergen3107)

**Review by mergen3107 — COMMENTED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/11277)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @mergen3107)

**Comment by mergen3107** [comment]
(sorry, was playing around with Reviewable. Don't know why it still says it is waiting on me)

**Comment by mergen3107** [comment]
Ready to merge?

**Comment by NiLuJe** [comment]
Assuming back is still the top-button, they ought to, yeah:

https://github.com/koreader/koreader/blob/63329569eb8afb00a29a723f215c6c5fcc9041aa/frontend/device/input.lua#L238-L246

**Comment by mergen3107** [comment]
Thanks! I hope these rotations will always yield the same results (Back/Forward on Top/Bottom) when launched in U/D.

**Comment by mergen3107** [comment]
Is going into 2023.12 release?

**Comment by hedger** [comment]
Tested on Oasis 3. Works fine, no more rotation/mirroring issues on startup.

**Review by hedger — APPROVED** [review]

**Comment by mergen3107** [comment]
@hedger Thank you very much!

Do buttons rotate properly as well in all orientations?

We are still trying to solve other issues that came up with this though :D 

If you start KOReader in any orientation, then flip it to the opposite one and restart KOReader via "Restart KOReader" in the top menu - KOReader will show up in that first orientation again (so, upside down after restart). 

It doesn't affect things much, touch and sensor are good.

**Comment by mergen3107** [comment]
See https://github.com/koreader/koreader/issues/11269#issuecomment-1876197092 and below

**Comment by NiLuJe** [comment]
> > Is going into 2023.12 release?
> 
> @NiLuJe ?

Probably not until we have a better grasp on #11269?

**Comment by NiLuJe** [comment]
Then again, the existing code doesn't make much sense, and this does, so, eeeh.

**Comment by mergen3107** [comment]
> Do buttons rotate properly as well in all orientations?

The answer is yes, a user from 4PDA confirmed that.

**Pull request merged**